streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java

```diff
@@ -648,71 +794,142 @@ public GroupAssignment assign(final Cluster metadata, final GroupSubscription gr
+            if (consumer == null) {
+                rebalanceRequired = true;
+                return interleaveTasksByGroupId(state.activeTasks(), consumers);
+            }
 
-        // assign previously assigned tasks to "old consumers"
-        for (final ClientMetadata clientMetadata : clientsMetadata.values()) {
-            for (final String consumerId : clientMetadata.consumers) {
+            // if this is a new task, or its old consumer no longer exists, it can be freely assigned
+            if (consumer.equals("")) {
+                newTasks.add(task);
+            } else {
+                assignments.computeIfAbsent(consumer, c -> new ArrayList<>());
 
-                if (futureConsumers.contains(consumerId)) {
-                    continue;
+                // if this task was owned by a consumer that is at capacity, abort
+                if (assignments.get(consumer).size() >= maxTasksPerClient) {
+                    rebalanceRequired = true;
+                    return interleaveTasksByGroupId(state.activeTasks(), consumers);
+                } else {
+                    assignments.get(consumer).add(task);
+                    // if we have now reach capacity, remove from set of consumers who need more tasks
+                    if (assignments.get(consumer).size() == maxTasksPerClient) {
+                        unfilledConsumers.remove(consumer);
+                    }
                 }
+            }
+        }
 
-                final List<TaskId> activeTasks = new ArrayList<>(clientMetadata.state.prevActiveTasks());
+        // initialize task list for any consumers not yet in the assignment
+        for (final String consumer : unfilledConsumers) {
+            assignments.computeIfAbsent(consumer, c -> new ArrayList<>());
+        }
 
-                final List<TopicPartition> assignedPartitions = new ArrayList<>();
-                for (final TaskId taskId : activeTasks) {
-                    assignedPartitions.addAll(partitionsForTask.get(taskId));
-                }
+        // interleave any remaining tasks among the remaining consumers with capacity
+        Collections.sort(newTasks);
+        while (!newTasks.isEmpty()) {
+            if (unfilledConsumers.isEmpty()) {
+                throw new IllegalStateException("Some tasks could not be distributed");
+            }
 
-                final Map<TaskId, Set<TopicPartition>> standbyTasks = new HashMap<>();
-                for (final TaskId taskId : clientMetadata.state.prevStandbyTasks()) {
-                    standbyTasks.put(taskId, partitionsForTask.get(taskId));
+            final Iterator<String> consumerIt = unfilledConsumers.iterator();
+            while (consumerIt.hasNext()) {
+                final String consumer = consumerIt.next();
+                final List<TaskId> consumerAssignment = assignments.get(consumer);
+                final TaskId task = newTasks.poll();
+                if (task == null) {
```

Comment:
Why this can happen? We checked `!newTasks.isEmpty()` before already right?

Reply:
We have a loop inside the main `while (!newTasks.isEmpty())` loop, where we are actively removing things from `newTasks`. So, we might have polled the last task in the `while (consumerIt.hasNext())` loop before we get back to the outer loop and check `newTasks.isEmpty`
This code is actually pretty much the same code as in the `interleaveTasksByGroupId` method. I'll see if it can be moved out into a shared method.